### PR TITLE
Add LoRA-GGPO for Flux

### PR DESCRIPTION
**Pull request by rockerBOO**
https://arxiv.org/abs/2502.14538v1

LoRA-GGPO (Gradient-Guided Perturbation Optimization) suggests they can mitigate double descent in LoRA. 

> The double descent phenomenon is a common non-
monotonic behavior in machine learning, where
model performance exhibits an “increase-decrease-
increase” trend with complexity or training time.

To do this they get the weight norm and gradient norm and use them to perturbed the activation's with a random matrix. 

| Ablation Study on GLUE Benchmark | Llama-2-7B results |
|--------|--------|
| Screenshot 2025-03-06 at 00-06-20 2502 14538v1 pdf | Screenshot 2025-03-06 at 00-06-37 2502 14538v1 pdf | 

Downside is it can increase training time by a semi-significant margin currently. I was able to get it down into the 20% decreased training speed but maybe better code could make it faster. In the paper they suggest 5% but I was not able to hit this.

To improve speed:
- Reduce the modules trained, each module will increase the training time
- Only train on some modules like attention and not on feed forward will be a little faster
- Skip steps on updating norms. Right now about 1.5s to update all the norms so not doing it on every step can save some time. It is set to update every 5 steps but we could make that adjustable. 

Usage:

`--network_args ggpo_sigma=0.03 ggpo_beta=0.01`

```
network_args = [
   "ggpo_sigma=0.03",
   "ggpo_beta=0.01"
]
```

>  σ(sigma) > 0 controls the
overall strength of the perturbation, with larger σ in-
creasing the perturbation range; β(beta) > 0 balances the
contributions of weight norms and gradient norms,
with larger β focusing more on gradient sensitiv-
ity and smaller β emphasizing weight importance

Generally keeping it as the values above would be good and all they represented in the paper. 

It is working at this stage and I have been using it. Originally implemented for the new Lumina LoRA but could be translated to SD1.5, SDXL LoRA's as well.

### CODE REVIEW

**Comment by feffy380** [comment]
Do you have any visual examples?

**Comment by rockerBOO** [comment]
Same settings but set  `ggpo_beta=0.01`, `ggpo_sigma=0.03` and thus using GGPO. Using my local version which I was able to improve performance to down to negligible performance hit. Will be getting a new version up soon. 
 

| Base | Non-GGPO | GGPO |
|--------|--------|--------|
| comfy_00003_ | comfy_00002_ | comfy_00001_| 

| Base | Non-GGPO | GGPO |
|--------|--------|--------|
| comfy_00005_ | comfy_00004_ | comfy_00006_ | 

| Base | Non-GGPO | GGPO |
|--------|--------|--------|
| comfy_00016_ | comfy_00015_ | comfy_00017_ | 

| Base | Non-GGPO | GGPO |
|--------|--------|--------|
| comfy_00031_ | comfy_00032_ | comfy_00030_ | 

| Base | Non-GGPO | GGPO |
|--------|--------|--------|
| comfy_00037_ | comfy_00038_ | comfy_00036_ |

| Base | Non-GGPO | GGPO |
|--------|--------|--------|
| comfy_00043_ | comfy_00044_ | comfy_00042_ |

| Base | Non-GGPO | GGPO |
|--------|--------|--------|
| comfy_00046_ | comfy_00045_ | comfy_00047_ |

**Comment by MegaCocos** [comment]
> Same settings but set `ggpo_beta=0.01`, `ggpo_sigma=0.03` and thus using GGPO.
> 
>     * 544 steps
> 
>     * train_double_block_indices=0,1,17,18
> 
>     * train_single_block_indices=34,35,37
> 
> 
> Using my local version which I was able to improve performance to down to negligible performance hit. Will be getting a new version up soon.


If I want to try LoRA training with GGPO function i guess i need to replace some files of sd-scripts with ones of your fork https://github.com/rockerBOO/sd-scripts/tree/lora-ggpo and use the command line option `--network_args ggpo_sigma=0.03 ggpo_beta=0.01` ? Could you tell me which files i need to replace? Then i see that you only train some blocks of FLUX, not all of them. Is it valid/sufficient/benefical for any kind of FLUX LoRA or it is only because of your specific LoRA kind? ¿Could you help me to understand the selection of blocks to training?

**Comment by rockerBOO** [comment]
If you check out that branch it is aligned with sd3 currently so should just work. The current branch is a bit slower than what I might have up in a few days but works the same way. 

Yes those parameters are all you would need (ggpo_beta, ggpo_sigma). You can confirm in the logs it will say
 `INFO     LoRA-GGPO training sigma: 0.03 beta: 0.01`

The train blocks was for clarification of the training and not needed or relevant. Mostly that it's not training all the blocks in the example, for clarification. 

It works only with `--network_module networks.lora_flux` version but anything under this should work fine.

**Comment by MegaCocos** [comment]
> If you check out that branch it is aligned with sd3 currently so should just work. The current branch is a bit slower than what I might have up in a few days but works the same way.
> 
> Yes those parameters are all you would need (ggpo_beta, ggpo_sigma). You can confirm in the logs it will say `INFO LoRA-GGPO training sigma: 0.03 beta: 0.01`
> 
> The train blocks was for clarification of the training and not needed or relevant. Mostly that it's not training all the blocks in the example, for clarification.
> 
> It works only with `--network_module networks.lora_flux` version but anything under this should work fine.

Thank you, i will try it! Suppose there is no any visual comparison still of how different values of sigma and beta affect the result. But i like a lot the hands consistence in the last example (the buffoon one), hope it could help in my case when the LoRA still not seems overtrained but the hands are already gone. My training is very large of 85000 steps and when finally the LoRA knows all i want the hands are similar to NON-GGPO buffoon :( Thank you for the clarification.

**Comment by rockerBOO** [comment]
Yeah the beta/sigma are from LLM training so the values for diffusion models may be different. It's a perturbation strength so maybe up to 0.1 could be a range. I might do some examples using some other values, but we'll see. 

Also the training may be overall faster because it is possibly controlling the  double descent, and helping with alignment. In my training above I have a few other changes that might make it more stable and perform a little better. I implemented debiased estimation and huber snr loss for flux which seems to perform really well. I will be making those be separate PR's in a few days as well.

**Comment by MegaCocos** [comment]
> Yeah the beta/sigma are from LLM training so the values for diffusion models may be different. It's a perturbation strength so maybe up to 0.1 could be a range. I might do some examples using some other values, but we'll see.
> 
> Also the training may be overall faster because it is possibly controlling the double descent, and helping with alignment. In my training above I have a few other changes that might make it more stable and perform a little better. I implemented debiased estimation, huber snr loss, and ip noise gamma for flux which seems to perform really well. I will be making those be separate PR's in a few days as well.

If you could share the command for implementation of all of these options for sd-scripts i would be very grateful:) I am stuck with this LoRA tree months already:)

**Comment by rockerBOO** [comment]
Some results showing different beta and sigma values. Seems like they can have a larger impact but beyond me to create any true assessment of what is better or worse. Would imagine higher values would create more "regularization" like effects due to the perturbation being higher. Sigma having more impact than beta. 

| beta 0.05, sigma 0.03 | beta 0.1, sigma 0.03 | beta 0.01, sigma 0.1 |
|--------|--------|--------|
| comfy_00105_ | comfy_00106_ | comfy_00107_ |

| Base | beta 0.05, sigma 0.03 | beta 0.1, sigma 0.03 | beta 0.01, sigma 0.1 |
|--------|--------|--------|--------|
| comfy_00114_ | comfy_00110_ | comfy_00109_ | comfy_00108_ |

| Base |  beta 0.05, sigma 0.03 | beta 0.1, sigma 0.03  | beta 0.01, sigma 0.1  |
|--------|--------|--------|--------|
| comfy_00135_ | comfy_00138_ | comfy_00137_ | comfy_00136_ |

| Base  |   beta 0.05, sigma 0.03 | beta 0.1, sigma 0.03  | beta 0.01, sigma 0.1  | 
|--------|--------|--------|--------|
|comfy_00148_ | comfy_00145_ | comfy_00146_ | comfy_00147_   

| Base  |   beta 0.05, sigma 0.03 | beta 0.1, sigma 0.03  | beta 0.01, sigma 0.1  | 
|--------|--------|--------|--------|
| comfy_00224_ | comfy_00220_ | comfy_00219_ | comfy_00218_ |

**Comment by rockerBOO** [comment]
| Base 	| beta 0.05, sigma 0.03 	| beta 0.1, sigma 0.03 	| beta 0.01, sigma 0.1|
|--------|--------|--------|--------|
| comfy_00101_ | comfy_00097_ | comfy_00096_ | comfy_00095_ | 

| Base 	| beta 0.05, sigma 0.03 	| beta 0.1, sigma 0.03 	| beta 0.01, sigma 0.1|
|--------|--------|--------|--------|
| comfy_00155_ |comfy_00151_ | comfy_00150_ | comfy_00149_ |

| Base 	| beta 0.05, sigma 0.03 	| beta 0.1, sigma 0.03 	| beta 0.01, sigma 0.1|
|--------|--------|--------|--------|
| comfy_00134_ | comfy_00131_ | comfy_00132_ | comfy_00133_ |

**Comment by rockerBOO** [comment]
Updated the norm process to make a norm estimation of the base weights on initialization. Then we don't have to move the weights around devices if using block swapping and also keep a reference around to the base weights. Error for the estimated norm is around 0.5% to 1.5% error. There is a validation function that can be used to check the error from the actual to the estimated norms. Maybe my approach is not completely accurate but because it's a random perturbation based on the gradient norms mostly, it might be ok. 

The norm is being put into the progress bar for all the norms but it might be too much info there. Also the norms are being logged which seems to be a good idea. 

This seems to be working as expected now.

File: networks/lora_flux.py [hunk 1]
```diff
@@ -140,7 +192,17 @@ def forward(self, x):
 
             lx = self.lora_up(lx)
 
-            return org_forwarded + lx * self.multiplier * scale
+            # LoRA Gradient-Guided Perturbation Optimization
+            if self.training and hasattr(self, 'perturbation_seed') and self.ggpo_sigma is not None and self.ggpo_beta is not None and self.combined_weight_norms is not None and self.grad_norms is not None:
```
[the remark this answers is not in the document]
Comment (rockerBOO) [reply networks/lora_flux.py:196]: the length of this if statement shows when other things were required/added. So yes we can remove this. By removing the temp random seed I think we can remove perturbation seed all together and rely on the rand seeding from the training.

**Comment by rockerBOO** [comment]
Only last bit is the progress bar probably has too much info in it at the moment (combined, key and grad norms). They are recorded in the logs so not necessary there.

**Comment by rockerBOO** [comment]
Not conclusive as the sample size is still small but tried a higher beta/sigma and seemed to work a little smoother in the results.

```
  "ggpo_beta=0.025",
  "ggpo_sigma=0.05",
 ```

of the gray bar that is longer. Not all the other graphs are similar but showcase the jumpy up and down of the loss curves. Not always indicative of something not working right or directly of double descent but this PR is suppose to be helping with that effect. 

Would **play with these values** to see which works better for **your dataset**, and should see some results somewhat early in the training (if it starts to go up and down more). 

 Screenshot 2025-03-27 at 16-34-38 women-flux-kohya-lora Workspace – Weights   Biases Screenshot 2025-03-27 at 16-33-43 valiant-sunset-240 women-flux-kohya-lora – Weights   Biases

**Comment by FurkanGozukara** [comment]
I am following this post since first post but what do we actually see here?

I mean

Base | Non-GGPO | GGPO

what is base training dataset?

**Comment by rockerBOO** [comment]
> what is base training dataset?

Base is the base Flux.1 dev model. Non-GGPO and GGPO are the trained LoRAs. Trained only a limited number of steps but on photos at 512x512 for testing purposes.

**Comment by FurkanGozukara** [comment]
> > what is base training dataset?
> 
> Base is the base Flux.1 dev model. Non-GGPO and GGPO are the trained LoRAs. Trained only a limited number of steps but on photos at 512x512 for testing purposes.

thank you. so you didnt compare with regular LoRA

**Comment by rockerBOO** [comment]
> thank you. so you didnt compare with regular LoRA

The Non-GGPO would be a regular LoRA (without the GGPO options)
